### PR TITLE
[Builder] Enhance builder to support allo.grid() with Non-Constant variables

### PR DESCRIPTION
**Pull request by AdrianLiu00**
## Description ##
Enhance builder to support allo.grid() with Non-Constant variables.

### Problems ###
The builder resolved bound variables as constants when building MLIR for allo.grid(). This led to parsing error if the bound variables are not constants, such as some named vairables with expression.

### Proposed Solutions ###
Enhance the for-loops building for grid() in builder with iterative construction through each axis and distinguished ForOps for two type of bound variables. 

### Examples ###
```
def test_grid_scf():
    def case1[A: int32, B: int32](X: int32) -> int32:
        T1: int32 = A + B
        T2: int32 = A - B
        for i, j in allo.grid(T1, T2):
            X += 1
        return X

    s = allo.customize(case1, instantiate=[3, 2])
    print(s.module)
```

```
#map = affine_map<() -> (0)>
"builtin.module"() ({
  "func.func"() <{function_type = (i32) -> i32, sym_name = "case1"}> ({
  ^bb0(%arg0: i32):
    %0 = "arith.constant"() <{value = 3 : i32}> : () -> i32
    %1 = "arith.extsi"(%0) : (i32) -> i33
    %2 = "arith.constant"() <{value = 2 : i32}> : () -> i32
    %3 = "arith.extsi"(%2) : (i32) -> i33
    %4 = "arith.addi"(%1, %3) : (i33, i33) -> i33
    %5 = "arith.trunci"(%4) : (i33) -> i32
    %6 = "memref.alloc"() <{odsOperandSegmentSizes = array<i32: 0, 0>}> {name = "T1"} : () -> memref<1xi32>
    "affine.store"(%5, %6) {map = #map, to = "T1"} : (i32, memref<1xi32>) -> ()
    %7 = "arith.constant"() <{value = 3 : i32}> : () -> i32
    %8 = "arith.extsi"(%7) : (i32) -> i33
    %9 = "arith.constant"() <{value = 2 : i32}> : () -> i32
    %10 = "arith.extsi"(%9) : (i32) -> i33
    %11 = "arith.subi"(%8, %10) : (i33, i33) -> i33
    %12 = "arith.trunci"(%11) : (i33) -> i32
    %13 = "memref.alloc"() <{odsOperandSegmentSizes = array<i32: 0, 0>}> {name = "T2"} : () -> memref<1xi32>
    "affine.store"(%12, %13) {map = #map, to = "T2"} : (i32, memref<1xi32>) -> ()
    %14 = "arith.constant"() <{value = 0 : i32}> : () -> i32
    %15 = "arith.index_cast"(%14) : (i32) -> index
    %16 = "affine.load"(%6) {from = "T1", map = #map} : (memref<1xi32>) -> i32
    %17 = "arith.index_cast"(%16) : (i32) -> index
    %18 = "arith.constant"() <{value = 1 : i32}> : () -> i32
    %19 = "arith.index_cast"(%18) : (i32) -> index
    "scf.for"(%15, %17, %19) ({
    ^bb0(%arg1: index):
      "scf.for"(%21, %23, %25) ({
      ^bb0(%arg2: index):
        %26 = "arith.constant"() <{value = 1 : i32}> : () -> i32
        %27 = "arith.addi"(%arg0, %26) : (i32, i32) -> i32
        "affine.store"(%27, %arg0) {map = #map, to = "X"} : (i32, i32) -> ()
        "scf.yield"() : () -> ()
      }) {loop_name = "j", op_name = "S_i_j_0"} : (index, index, index) -> ()
      "scf.yield"() : () -> ()
    }) {loop_name = "i", op_name = "S_i_j_0"} : (index, index, index) -> ()
    %20 = "arith.constant"() <{value = 0 : i32}> : () -> i32
    %21 = "arith.index_cast"(%20) : (i32) -> index
    %22 = "affine.load"(%13) {from = "T2", map = #map} : (memref<1xi32>) -> i32
    %23 = "arith.index_cast"(%22) : (i32) -> index
    %24 = "arith.constant"() <{value = 1 : i32}> : () -> i32
    %25 = "arith.index_cast"(%24) : (i32) -> index
    "func.return"(%arg0) : (i32) -> ()
  }) {itypes = "s", otypes = "s"} : () -> ()
}) : () -> ()
```

## Checklist ##

- [x] PR's title starts with a category (e.g. [Bugfix], [IR], [Builder], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage (It would be better to provide ~2 different test cases to test the robustness of your code)
- [x] Code is well-documented

### CODE REVIEW

File: allo/ir/builder.py [hunk 1]
```diff
@@ -201,8 +201,74 @@ def build_all_for(ctx, node, attr):
...
         iter_args = node.iter.args
         if attr in {"grid", "reduction"}:
-            grid = [ASTResolver.resolve_constant(x, ctx) for x in iter_args]
-            for_loops = build_for_loops(grid, ctx.get_ip(), names, stage_name)
+            for_loops = []
+            if stage_name is None:
+                stage_name = "S_" + "_".join(names)
+
+            ip_handle = ctx.get_ip()
+            for ind, arg in enumerate(iter_args):  # Traversal
```
Comment (chhzh123) [inline allo/ir/builder.py:286]: Did you copy the following code from the other branch? Maybe create a common function to reduce duplicated code

File: tests/test_schedule_compute.py [hunk 2]
```diff
@@ -22,6 +22,18 @@ def gemm(A: int32[32, 32], B: int32[32, 32]) -> int32[32, 32]:
...
+def test_grid_scf():
+    def case1[A: int32, B: int32](X: int32) -> int32:
+        T1: int32 = A + B
+        T2: int32 = A - B
+        for i, j in allo.grid(T1, T2):
+            X += 1
+        return X
+
+    s = allo.customize(case1, instantiate=[3, 2])
+    print(s.module)
```
Comment (chhzh123) [inline tests/test_schedule_compute.py]: Need to add NumPy tests

**Comment by AdrianLiu00** [comment]
okay, I will check it out and do further modification.

**Comment by AdrianLiu00** [comment]
I have fixed aforementioned issues and organized code in a more concise way. Please check it out if there still needs some enhancement. By the way, the function `build_for_loops()` in [transform.py](https://github.com/cornell-zhang/allo/blob/main/allo/ir/transform.py) is no longed used in builder. If there no other parts use this function, it can be removed as well.

**Review by chhzh123 — APPROVED** [review]
LGTM. This PR is helpful in the way that unifies the grid- and range-for builder process. Thanks!

**Pull request merged**